artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/sasl/MechanismFinder.java

```diff
@@ -17,11 +17,31 @@
 
 package org.apache.activemq.artemis.protocol.amqp.sasl;
 
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.ServiceLoader;
+import java.util.stream.Stream;
+
 public class MechanismFinder {
 
-   public static String[] KNOWN_MECHANISMS = new String[]{PlainSASL.NAME, AnonymousServerSASL.NAME};
+   private static final String[] DEFAULT_MECHANISMS = new String[] {PlainSASL.NAME, AnonymousServerSASL.NAME};
+
+   private static final Map<String, ServerSASLFactory> FACTORY_MAP = new HashMap<>();
+
+   static {
+      ServiceLoader<ServerSASLFactory> serviceLoader =
+               ServiceLoader.load(ServerSASLFactory.class, MechanismFinder.class.getClassLoader());
+      for (ServerSASLFactory factory : serviceLoader) {
+         FACTORY_MAP.put(factory.getMechanism(), factory);
+      }
+   }
 
    public static String[] getKnownMechanisms() {
-      return KNOWN_MECHANISMS;
+      return Stream.concat(Arrays.stream(DEFAULT_MECHANISMS), FACTORY_MAP.keySet().stream()).toArray(String[]::new);
```

Comment:
Thanks for the advice, I have added now a "precedence" that is used to maintains it ordering. In case of duplicate mechanism registered the one with the highest precedence wins.

Reply:
With the precedence feature, having the existing mechanisms use this loader makes sense. The actual published mechanisms on the amqp acceptor can be explicitly configured on amqp protocol manager via the saslMechanisms property. That separates the loading from the use, but the defaults should have a sensible order as @gemmellr  points out. I think there are existing tests that verify that.

Reply:
maybe use "position", follow java.security.Security#insertProviderAt - a provider at an existing position cannot be overridden. In this way it can be clear what the default order is. duplicates need to figure out where they fit in the ordered list to be effective.